File: ImasLiveDB/Services/SetlistLikeService.swift
import Foundation

/// 公演 (show) ごとの「この曲良かった」 like 状態をサーバから取得 + 投票するクライアント。
/// star toggle なので idempotent: 多重 POST しても 1 票、 DELETE は 0 票まで。
@Observable @MainActor
final class SetlistLikeService {
    static let shared = SetlistLikeService()
    private init() {}

    struct LikeEntry: Codable, Sendable {
        let songId: String
        let likeCount: Int
        let hasUserLiked: Bool
    }

    struct LikeResult: Codable, Sendable {
        let songId: String
        let likeCount: Int
        let liked: Bool
    }

    /// 公演ごとの like 集計 (/shows/:id/likes) の TTL キャッシュ。show_id 単位。
    /// レスポンスは has_user_liked (自分の like か) を含む **ユーザー固有** データなので、
    /// エッジ/CDN の共有キャッシュには絶対載せない (backend も public を付けない)。
    /// per-device のメモリキャッシュのみ。同じ公演のセトリを開き直すたびに叩くのを抑える。
    /// like 数は他人の操作でも増減するので TTL は短め (60s)。
    /// 自分の like/unlike は該当 song のエントリをその場で patch して整合を保つ。
    private var likesCache: [String: (entries: [LikeEntry], at: Date)] = [:]
    private let likesCacheTTL: TimeInterval = 60

    func fetch(showId: String) async throws -> [LikeEntry] {
        if let hit = likesCache[showId], Date().timeIntervalSince(hit.at) < likesCacheTTL {
            return hit.entries
        }
        let entries: [LikeEntry] = try await APIClient.shared.request(
            "GET",
            path: "/shows/\(showId)/likes",
            authorized: true
        )
        likesCache[showId] = (entries, Date())
        return entries
    }

    // bearerToken の事前チェックはしない。期限切れ (nil) でも isSignedIn は true のまま
    // 残ることがあり、ここで弾くと APIClient の 401 自動リフレッシュに乗らず Good が無言で
    // 失敗する (セトリ予想と同じ症状)。authorized: true で送れば未トークン→401→refresh→retry。
    func like(showId: String, songId: String) async throws -> LikeResult {
        let result: LikeResult = try await APIClient.shared.request(
            "POST",
            path: "/shows/\(showId)/songs/\(songId)/like",
            authorized: true
        )
        applyLikeResult(showId: showId, result: result)
        return result
    }

    func unlike(showId: String, songId: String) async throws -> LikeResult {
        let result: LikeResult = try await APIClient.shared.request(
            "DELETE",
            path: "/shows/\(showId)/songs/\(songId)/like",
            authorized: true
        )
        applyLikeResult(showId: showId, result: result)
        return result
    }

    /// 自分の like/unlike の結果をキャッシュに反映する。サーバが返した最新の count/liked を
    /// その曲のエントリに patch するので、古い has_user_liked が残らない。
    private func applyLikeResult(showId: String, result: LikeResult) {
        guard let cached = likesCache[showId] else { return }
        let entry = LikeEntry(songId: result.songId, likeCount: result.likeCount, hasUserLiked: result.liked)
        var entries = cached.entries
        if let idx = entries.firstIndex(where: { $0.songId == result.songId }) {
            entries[idx] = entry
        } else {
            entries.append(entry)
        }
        // 集計の鮮度は維持しつつ自分の操作だけ反映 (at は据え置きで TTL を延ばさない)。
        likesCache[showId] = (entries, cached.at)
    }

    /// サインアウト/アカウント切替時に呼ぶ。has_user_liked は user 依存なので、
    /// 別ユーザーの like 状態が残らないよう全公演の集計キャッシュを捨てる。
    func clearCache() {
        likesCache.removeAll()
    }
}

enum LikeError: LocalizedError {
    case unauthorized
    var errorDescription: String? {
        switch self {
        case .unauthorized: return "Like するには Apple Sign In が必要です"
        }
    }
}

File: ImasLiveDB-Android/app/src/main/kotlin/com/fugaif/imaslivedb/ui/events/SetlistLikeService.kt
package com.fugaif.imaslivedb.ui.events

import android.content.Context
import android.util.Log
import com.fugaif.imaslivedb.data.community.DeviceIdentity
import com.fugaif.imaslivedb.di.AppModule
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

/**
 * 公演ごとの「この曲良かった」 like をサーバから取得 / 投票するクライアント。
 * iOS `SetlistLikeService` の移植。
 *
 * star toggle なので idempotent — 多重 POST しても 1 票、DELETE は 0 票まで。
 *
 * 集計レスポンスは `has_user_liked` (自分の like か) を含む**ユーザー固有**データなので、
 * 共有キャッシュには絶対載せない。ここが持つのは端末内メモリの短命 TTL キャッシュだけで、
 * 同じ公演のセトリを開き直すたびにサーバを叩くのを抑えるためのもの。like 数は他人の操作でも
 * 増減するので TTL は短く (60 秒)、自分の like/unlike は該当曲だけその場で patch する。
 */
class SetlistLikeService private constructor(private val appContext: Context) {

    /** 1 曲ぶんの集計 + 自分の like 状態。 */
    data class LikeEntry(val songId: String, val likeCount: Int, val hasUserLiked: Boolean)

    /** like / unlike が「認証されていない」で弾かれたことを表す。呼び出し側はログイン誘導に使う。 */
    class Unauthorized : Exception("Like するにはログインが必要です")

    private class CacheHit(val entries: List<LikeEntry>, val atMillis: Long)

    private val cache = HashMap<String, CacheHit>()

    /** 公演の全曲ぶんの集計。TTL 内なら再取得しない。 */
    suspend fun fetch(showId: String): List<LikeEntry> = withContext(Dispatchers.IO) {
        synchronized(cache) {
            cache[showId]?.takeIf { System.currentTimeMillis() - it.atMillis < CACHE_TTL_MS }
        }?.let { return@withContext it.entries }

        val (code, body) = request("GET", "/shows/${enc(showId)}/likes")
        if (code !in 200..299 || body.isNullOrEmpty()) return@withContext emptyList()
        val entries = runCatching {
            val arr = JSONArray(body)
            (0 until arr.length()).map { i ->
                val o = arr.getJSONObject(i)
                LikeEntry(o.optString("song_id"), o.optInt("like_count"), o.optBoolean("has_user_liked"))
            }
        }.getOrDefault(emptyList())
        synchronized(cache) { cache[showId] = CacheHit(entries, System.currentTimeMillis()) }
        entries
    }

    suspend fun like(showId: String, songId: String): LikeEntry =
        toggle("POST", showId, songId)

    suspend fun unlike(showId: String, songId: String): LikeEntry =
        toggle("DELETE", showId, songId)

    private suspend fun toggle(method: String, showId: String, songId: String): LikeEntry =
        withContext(Dispatchers.IO) {
            val (code, body) = request(method, "/shows/${enc(showId)}/songs/${enc(songId)}/like")
            // 401 は「トークンが無い / 失効」。ここで isSignedIn を先読みして弾かないのは、
            // セッション更新中の一瞬でも投票を無言で落としたくないため (iOS と同じ判断)。
            if (code == 401 || code == 403) throw Unauthorized()
            if (code !in 200..299 || body.isNullOrEmpty()) throw IllegalStateException("like failed: HTTP $code")
            val json = JSONObject(body)
            val entry = LikeEntry(
                songId = json.optString("song_id").ifEmpty { songId },
                likeCount = json.optInt("like_count"),
                hasUserLiked = json.optBoolean("liked")
            )
            patchCache(showId, entry)
            entry
        }

    /**
     * 自分の操作の結果だけキャッシュに反映する。取得時刻 (atMillis) は据え置きで、
     * 集計そのものの鮮度 (他人の票) は延ばさない。
     */
    private fun patchCache(showId: String, entry: LikeEntry) {
        synchronized(cache) {
            val hit = cache[showId] ?: return
            val entries = hit.entries.filter { it.songId != entry.songId } + entry
            cache[showId] = CacheHit(entries, hit.atMillis)
        }
    }

    /** サインアウト / アカウント切替時に呼ぶ。has_user_liked は利用者ごとの値なので全部捨てる。 */
    fun clearCache() {
        synchronized(cache) { cache.clear() }
    }

    private fun enc(s: String): String = URLEncoder.encode(s, "UTF-8").replace("+", "%20")

    /** ステータスとレスポンス本文を返す最小の HTTP。通信自体が失敗したら code = -1。 */
    private fun request(method: String, path: String): Pair<Int, String?> = try {
        val conn = (URL(BASE + path).openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 15_000
            readTimeout = 15_000
            setRequestProperty("Content-Type", "application/json")
            setRequestProperty("X-Device-Id", DeviceIdentity.get(appContext))
            AppModule.from(appContext).authService.sessionToken
                ?.let { setRequestProperty("Authorization", "Bearer $it") }
        }
        val code = conn.responseCode
        val text = (if (code in 200..299) conn.inputStream else conn.errorStream)
            ?.bufferedReader()?.use { it.readText() }
        conn.disconnect()
        code to text
    } catch (e: Exception) {
        Log.w(TAG, "$method $path failed: ${e.message}")
        -1 to null
    }

    companion object {
        private const val BASE = "https://imas-live-api.tokata3011.workers.dev"
        private const val TAG = "SetlistLike"
        private const val CACHE_TTL_MS = 60_000L

        @Volatile
        private var instance: SetlistLikeService? = null

        /** 集計キャッシュを画面をまたいで共有するため単一インスタンスにする。 */
        fun get(context: Context): SetlistLikeService =
            instance ?: synchronized(this) {
                instance ?: SetlistLikeService(context.applicationContext).also { instance = it }
            }
    }
}
